docker/postgres-metadb/metadb_s3_tarz

```diff
@@ -0,0 +1,209 @@
+#!/usr/bin/env python3
+#
+# This script runs at metadb PostgreSQL instance as a separate
+# memory-constrained container launched by Airflow scheduler.
+#
+# It creates a base backup in S3 bucket https://ooni-data.s3.amazonaws.com/metadb/
+#
+# It expects AWS_* credentials as environment variables, `awscli` uses
+# AWS_DEFAULT_REGION, AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY.
+# PG* variables are interpreted by `psql` binary used to call `pg_start_backup()`.
+#
+# It may be replaced by barman (that does not upload to S3 right now) or other solution.
+#
+
+from contextlib import contextmanager
+import concurrent.futures
+import glob
+import gzip
+import logging
+import os
+import queue
+import signal
+import subprocess
+import sys
+import tempfile
+import time
+import urllib.request
+
+assert sys.version_info > (3, 5)
+
+for _ in ('AWS_ACCESS_KEY_ID', 'AWS_SECRET_ACCESS_KEY'):
+    assert os.environ.get(_), _
+
+assert os.path.isdir('base/pgsql_tmp') and os.access('base/pgsql_tmp', os.W_OK)
+TMPDIR = 'base/pgsql_tmp'
+TMPPFX = 's3tarz'
+
+with open('PG_VERSION') as _:
+    PGVER = _.read().strip()
+UNAME = os.uname()
+NOW = int(time.time())
+LATEST = '{:04d}{:02d}{:02d}_{:d}'.format(*(time.gmtime(NOW)[:3]), NOW)
+S3BASE = 'metadb/{}-{}-{}/base/'.format(UNAME.sysname, UNAME.machine, PGVER)
+S3DIR = S3BASE + LATEST + '/'
+S3_ROOT = 's3://ooni-data/'
+HTTP_ROOT = 'https://ooni-data.s3.amazonaws.com/'
+
+def pg_files():
+    def walk_onerror(arg):
+        raise arg # to die in case of EPERM or something like that
+    norecurse = {
+        # https://www.postgresql.org/docs/10/continuous-archiving.html#BACKUP-LOWLEVEL-BASE-DATA
+        'pg_dynshmem', 'pg_notify', 'pg_serial', 'pg_snapshots', 'pg_stat_tmp', 'pg_subtrans',
+        # master-only files
+        'pg_replslot',
+        # WAL for PG 10+ and PG<=9.6
+        'pg_wal', 'pg_xlog',
+        # server logs
+        'pg_log',
+    }
+    noroot = {'postmaster.pid', 'postmaster.opts'}
+    ret = []
+    for root, dirs, files in os.walk('.', onerror=walk_onerror):
+        for container in (dirs, files):
+            for fobj in container[:]:
+                if fobj.startswith('pgsql_tmp'):
+                    container.remove(fobj)
+        for fobj in sorted(dirs):
+            ret.append((os.path.join(root, fobj), 0)) # dirs are always created, but may be empty
+        if root == '.':
+            for fobj in set(dirs) & norecurse:
+                dirs.remove(fobj) # in-place update of the list disables recursion
+            for fobj in set(files) & noroot:
+                files.remove(fobj)
+        for fobj in sorted(files):
+            fpath = os.path.join(root, fobj)
+            ret.append((fpath, os.path.getsize(fpath)))
+    return ret
+
+def file_buckets(fpath_size):
+    ret = []
+    cur, cumsize = [], 0
+    for t in fpath_size:
+        _, size = t
+        cur.append(t)
+        cumsize += size
+        if cumsize > 1610612736: # 1.5GiB
```

Comment:
Maybe this should be defined as a constant in the beginning of the file to make it more obvious.